### PR TITLE
client: implemented messaging signing.

### PR DESCRIPTION
**Pull request by ardevd**
added support for the main RPC implementation of message signing.

### CODE REVIEW

File: lightning_client.go [hunk 1]
```diff
@@ -4132,6 +4139,20 @@ func (s *lightningClient) SendCustomMessage(ctx context.Context,
 	return err
 }
 
+// SignMessage signs a message with this node's private key.
+// The returned signature string is zbase32 encoded and pubkey
+// recoverable, meaning that only the message digest and signature
+// are needed for verification.
+func (s *lightningClient) SignMessage(ctx context.Context,
+	data []byte) (string, error) {
+	rpcCtx := s.adminMac.WithMacaroonAuth(ctx)
```
Comment (guggero) [inline lightning_client.go:4148]: nit: extra line between multi-line function header and first code.

File: lightning_client.go [hunk 2]
```diff
@@ -4132,6 +4139,20 @@ func (s *lightningClient) SendCustomMessage(ctx context.Context,
...
+
+	rpcReq := &lnrpc.SignMessageRequest{Msg: data}
+	rpcRes, err := s.client.SignMessage(rpcCtx, rpcReq)
```
Comment (guggero) [inline lightning_client.go:4151]: Need to add an `if err != nil ` check here. Otherwise `rpcRes.Signature` below will cause a nil pointer panic.

**Review by guggero — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: lightning_client.go (hunk 2 above)
Comment (ardevd) [reply lightning_client.go:4151]: Ah, nice catch. Should have caught that one.

File: lightning_client.go [hunk 3]
```diff
@@ -4145,10 +4145,13 @@ func (s *lightningClient) SendCustomMessage(ctx context.Context,
...
-	rpcCtx := s.adminMac.WithMacaroonAuth(ctx)
 
+	rpcCtx := s.adminMac.WithMacaroonAuth(ctx)
 	rpcReq := &lnrpc.SignMessageRequest{Msg: data}
 	rpcRes, err := s.client.SignMessage(rpcCtx, rpcReq)
+	if err != nil {
+		return "", err
+	}
 
 	return rpcRes.Signature, err
```
Comment (guggero) [inline lightning_client.go:4156]: Should now just return `rpcRes.Signature, nil` here.

Comment (ardevd) [reply lightning_client.go]: Sharp eyed reviewer! Fixed.

**Review by bhandras — APPROVED** [review]
I believe you need to update `permissions.json` to make the test pass.
Other than that LGTM 🎉

**Comment by ardevd** [comment]
> I believe you need to update `permissions.json` to make the test pass. Other than that LGTM 🎉

Not entirely sure what update is needed to `permissions.json`. Something like:

```json
"/lnrpc.Lightning/Signmessage": {
            "permissions": [
                {
                    "entity": "onchain",
                    "action": "read"
                }
            ]
        },
```
?

**Comment by guggero** [comment]
Yes, that should be enough.

**Comment by ardevd** [comment]
Actually, on further inspection, a similar block is already present in permissions.json.

```json
"/lnrpc.Lightning/SignMessage": {
            "permissions": [
                {
                    "entity": "message",
                    "action": "write"
                }
            ]
        },
```

on line 563

**Comment by guggero** [comment]
Ah, I think you need to bump Line 15 in https://github.com/lightninglabs/lndclient/blob/2005b0d5f6a33397a7679bb722ed096c77fe55c2/macaroon_recipes_test.go#L15.
Since a new permission is now being used by the `lndclient`.

**Pull request merged**